src/Uno.Toolkit.UI/Behaviors/ItemsRepeaterExtensions.cs

```diff
@@ -474,18 +513,54 @@ private static bool SetSelectionStates(ItemsRepeater ir, params int[]? indexes)
 			SetSelectedItem(ir, items[0]);
 			SetSelectedIndexes(ir, indexes);
 			SetSelectedItems(ir, items);
+			SetSelectionInfoRanges(ir, indexes, diffIndexes);
 		}
 		else
 		{
 			SetSelectedIndex(ir, -1);
 			SetSelectedItem(ir, null);
 			SetSelectedIndexes(ir, Array.Empty<int>());
 			SetSelectedItems(ir, Array.Empty<object>());
+			SetSelectionInfoRanges(ir, Array.Empty<int>());
+
 		}
 
 		return true;
 	}
 
+	private static void SetSelectionInfoRanges(ItemsRepeater ir, int[] selectedIndexes, int[]? diffIndexes = null)
+	{
+		if (ir.ItemsSource is not ISelectionInfo selectionInfo) return;
+
+		IEnumerable<int>? indexesToDeselect = null;
+
+		// diffIndexes is used to optimize the selection change, by deselecting only the elements that are not in the new selection.
+		if (diffIndexes is { Length: > 0 })
+		{
+			var selectionLookup = diffIndexes.ToLookup(keySelector: idx => selectedIndexes.Contains(idx));
+
+			indexesToDeselect = selectionLookup[false];
+			selectedIndexes = selectionLookup[true].ToArray();
+		}
+		else
+		{
+			var selectedRanges = selectionInfo.GetSelectedRanges();
+
+			indexesToDeselect = selectedRanges.SelectMany(ExpandRange)
+					.Where(toDeselect => !selectedIndexes.Contains(toDeselect));
+		}
```

Comment:
not optimizing the selection range?
selectedRanges vs selectedIndex would be most likely the same for multi-select, with a single element in difference needed to update